crates/router/src/core/locking.rs

```diff
@@ -0,0 +1,191 @@
+            expiry_in_seconds,
+            retries,
+            delay_between_retries_in_seconds,
+            api_identifier,
+        )
+        .await
+    } else {
+        logger::info!(
+            "Resource Locking not Enabled for merchant_id: {} and api: {}",
+            locking_input.merchant_id,
+            api_identifier.to_owned()
+        );
+        Ok(LockStatus::NotEnabled)
+    }
+}
+
+#[instrument(skip(state))]
+pub async fn lock_resource(
+    state: &routes::AppState,
+    locking_key: String,
+    expiry_in_seconds: u64,
+    retries: i32,
+    delay_between_retries_in_seconds: i64,
+    _api_identifier: &str,
+) -> RouterResult<LockStatus> {
+    let redis_key_for_lock = get_redis_key_for_locks(locking_key);
+
+    acquire_lock_on_resource_in_redis(
+        state,
+        redis_key_for_lock.as_str(),
+        true,
+        expiry_in_seconds,
+        delay_between_retries_in_seconds,
+        retries,
+    )
+    .await
+}
+
+#[inline]
+fn get_redis_key_for_locks(a: String) -> String {
+    format!("API_LOCK_{}", a)
+}
+
+#[instrument(skip(state))]
+pub async fn acquire_lock_on_resource_in_redis(
+    state: &routes::AppState,
+    key: &str,
+    value: bool,
+    expiry_in_seconds: u64,
+    _delay_between_retries_in_seconds: i64,
+    mut retries: i32,
+) -> RouterResult<LockStatus> {
+    let redis_conn = state.store.clone().get_redis_conn();
+
+    while retries != 0 {
+        retries -= 1;
+        let is_lock_acquired = redis_conn
+            .set_key_if_not_exists_with_expiry(
+                key,
+                value,
+                Some(
+                    expiry_in_seconds
+                        .try_into()
+                        .into_report()
+                        .change_context(errors::ApiErrorResponse::InternalServerError)?,
+                ),
+            )
+            .await;
+
+        match is_lock_acquired {
+            Ok(redis::SetnxReply::KeySet) => {
+                return Ok(LockStatus::Acquired(key.to_owned()));
+            }
+            Ok(redis::SetnxReply::KeyNotSet) => {
+                logger::error!("Lock not acquired, retrying");
+                actix_time::sleep(tokio::time::Duration::from_millis(expiry_in_seconds * 1000))
+                    .await;
+            }
+            Err(error) => {
+                logger::error!(error=%error.current_context(), "Error while locking");
+                actix_time::sleep(tokio::time::Duration::from_millis(expiry_in_seconds * 1000))
+                    .await;
+            }
+        }
+    }
+
+    Ok(LockStatus::AcquireFailedRetriesExhausted(key.to_owned()))
+}
+
+#[instrument(skip(state))]
+pub async fn release_lock(
+    state: &routes::AppState,
+    mut retries: i32,
+    lock: LockStatus,
+) -> RouterResult<LockStatus> {
+    let redis_conn = state.store.clone().get_redis_conn();
+
+    match lock {
+        LockStatus::Acquired(key) | LockStatus::AlreadyLocked(key) => {
+            while retries != 0 {
```

Comment:
This doesn't prevent negative cases, can we force `retries` variable to be `u32`?